core/arch/arm/kernel/thread_optee_smc.c

```diff
@@ -523,15 +523,23 @@ static struct mobj *get_rpc_alloc_res(struct optee_msg_arg *arg,
 {
 	struct mobj *mobj = NULL;
 	uint64_t cookie = 0;
+	size_t size = 0;
+	paddr_t pa = 0;
 
 	if (arg->ret || arg->num_params != 1)
 		return NULL;
 
 	if (arg->params[0].attr == OPTEE_MSG_ATTR_TYPE_TMEM_OUTPUT) {
+		pa = arg->params[0].u.tmem.buf_ptr;
+		size = arg->params[0].u.tmem.size;
 		cookie = arg->params[0].u.tmem.shm_ref;
-		mobj = mobj_shm_alloc(arg->params[0].u.tmem.buf_ptr,
-				      arg->params[0].u.tmem.size,
-				      cookie);
+
+		if (core_pbuf_is(CORE_MEM_NSEC_SHM, pa, size))
+			mobj = mobj_shm_alloc(pa, size, cookie);
+#ifdef CFG_CORE_DYN_SHM
+		else if ((!(pa & SMALL_PAGE_MASK)) && size <= SMALL_PAGE_SIZE)
```

Comment:
`&& IS_ENABLED(CFG_CORE_DYN_SHM)` instead of the `#ifdef`.

Reply:
When can this happen?

Reply:
Could it be that instead there's a mismatch between what OP-TEE considers NSEC_SHM and what's the kernels view?

Reply:
requesting cache from rpc_io_i2c_transfer

Reply:
> `&& IS_ENABLED(CFG_CORE_DYN_SHM)` instead of the `#ifdef`.

sure but that would be inconsistent with the same code on the same file.
should I fix that as well? just to be coherent.....

Reply:
> > `&& IS_ENABLED(CFG_CORE_DYN_SHM)` instead of the `#ifdef`.
> 
> sure but that would be inconsistent with the same code on the same file (check line 315)
> should I fix that as well? just to be coherent.....

Sure, sounds good to me.

Reply:
I don't understand the question. Can you provide some context or point me to the documentation you are referring to?

The use cases basically are:

- CFG_CORE_RESERVED_SHM = y
- CFG_CORE_DYN_SHM = y
==> rpc_io_i2c fails to allocate 

- CFG_CORE_RESERVED_SHM = y
- CFG_CORE_DYN_SHM = n
==> rpc_io_i2c allocates correctly

- CFG_CORE_RESERVED_SHM = y
- CFG_CORE_DYN_SHM = y
- this patch
==> rpc_io_i2c allocates correctly

if you could help me understand why the first case must not work I'd be glad to close the PR.

Reply:
> Could it be that instead there's a mismatch between what OP-TEE considers NSEC_SHM and what's the kernels view?

ah, maybe that is it

Reply:
also, if you are rejecting this patch, could you explain why shm_cache allocation should be treated differently to argument allocation for messages?

Reply:
We have an ABI which the kernel also is supposed to follow, right now it looks like it's not. I don't know why it doesn't follow the ABI, but I'm quite certain the best place to fix that isn't here.
https://github.com/OP-TEE/optee_os/blob/master/core/include/optee_msg.h#L45-L71

Reply:
interesting. 
now, why thread_rpc_alloc_arg and get_rpc_alloc_res behave differently when allocating shared memory?

Reply:
sounds great.